include/forcing/ForcingsEngineDataProvider.hpp

```diff
@@ -0,0 +1,211 @@
+#pragma once
+
+#include <memory>
+#include <unordered_map>
+#include <chrono>
+
+#include "DataProvider.hpp"
+#include "bmi/Bmi_Py_Adapter.hpp"
+
+namespace data_access {
+
+//! Python module name for NextGen Forcings Engine
+static constexpr auto forcings_engine_python_module = "NextGen_Forcings_Engine";
+
+//! Python classname for NextGen Forcings Engine BMI class
+static constexpr auto forcings_engine_python_class  = "NWMv3_Forcing_Engine_BMI_model";
+
+//! Full Python classpath for Forcings Engine BMI class
+static constexpr auto forcings_engine_python_classpath = "NextGen_Forcings_Engine.NWMv3_Forcing_Engine_BMI_model";
+
+//! Default time format used for parsing timestamp strings
+static constexpr auto default_time_format = "%Y-%m-%d %H:%M:%S";
+
+namespace detail {
+
+//! Parse time string from format.
+//! Utility function for ForcingsEngineLumpedDataProvider constructor.
+time_t parse_time(const std::string& time, const std::string& fmt);
+
+//! Check that requirements for running the forcings engine
+//! are available at runtime. If requirements are not available,
+//! then this function throws.
+void assert_forcings_engine_requirements();
+
+//! Storage for Forcings Engine-specific BMI instances.
+struct ForcingsEngineStorage {
+    using key_type   = std::string;
+    using bmi_type   = models::bmi::Bmi_Py_Adapter;
+    using value_type = std::shared_ptr<bmi_type>;
```

Comment:
It would be good to comment these types with descriptions of the particular contents of their instances -- i.e. config file path, and corresponding instantiations of the above-named Python class